### ISSUE TITLE
wti-predictor

### ISSUE THREAD

**Issue by Maik994** [apz33189793/Oil-Price-WTI-Predict-by-LSTM-.#2]
import streamlit as st
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import MinMaxScaler
from keras.models import Sequential
from keras.layers import Dense, LSTM

st.title("Previsione Giornaliera del Prezzo del WTI")

# Carica i dati
uploaded_file = st.file_uploader("Carica il file CSV con i dati storici del WTI", type="csv")

if uploaded_file is not None:
    df = pd.read_csv(uploaded_file)
    st.write("Anteprima dati:", df.tail())

    if 'Close' not in df.columns:
        st.error("Il file deve contenere una colonna 'Close'")
    else:
        data = df['Close'].values.reshape(-1, 1)

        scaler = MinMaxScaler(feature_range=(0, 1))
        scaled_data = scaler.fit_transform(data)

        sequence_length = 60
        x_train, y_train = [], []

        for i in range(sequence_length, len(scaled_data)):
            x_train.append(scaled_data[i - sequence_length:i, 0])
            y_train.append(scaled_data[i, 0])

        x_train, y_train = np.array(x_train), np.array(y_train)
        x_train = np.reshape(x_train, (x_train.shape[0], x_train.shape[1], 1))

        model = Sequential()
        model.add(LSTM(units=50, return_sequences=True, input_shape=(x_train.shape[1], 1)))
        model.add(LSTM(units=50))
        model.add(Dense(1))

        model.compile(optimizer='adam', loss='mean_squared_error')
        model.fit(x_train, y_train, epochs=10, batch_size=32, verbose=0)

        last_60_days = scaled_data[-60:]
        last_60_days = np.reshape(last_60_days, (1, last_60_days.shape[0], 1))
        prediction = model.predict(last_60_days)
        prediction_price = scaler.inverse_transform(prediction)

        st.success(f"Prezzo previsto per il giorno successivo: **{prediction_price[0][0]:.2f} USD**")

        st.subheader("Andamento storico del prezzo WTI")
        fig, ax = plt.subplots()
        ax.plot(df['Close'], label="Storico")
        ax.set_xlabel("Giorni")
        ax.set_ylabel("Prezzo")
        ax.legend()
        st.pyplot(fig)
